src/vm/assembly.cpp

```diff
@@ -1780,6 +1780,36 @@ INT32 Assembly::ExecuteMainMethod(PTRARRAYREF *stringArgs, BOOL waitForOtherThre
     INT32   iRetVal = 0;
 
     BEGIN_ENTRYPOINT_THROWS;
+    {
+        GCX_PREEMP();
+
+        mdToken tkEntryPoint = m_pManifestFile->GetEntryPointToken();
+        // <TODO>@TODO: What if the entrypoint is in another file of the assembly?</TODO>
+        ReleaseHolder<IMDInternalImport> scope(m_pManifestFile->GetMDImportWithRef());
+        // In theory, we should have a valid executable image and scope should never be NULL, but we've been  
+        // getting Watson failures for AVs here due to ISVs modifying image headers and some new OS loader 
+        // checks (see Dev10# 718530 and Windows 7# 615596)
+        if (scope == NULL)
+        {
+            ThrowHR(COR_E_BADIMAGEFORMAT);
+        }
+
+#ifdef FEATURE_COMINTEROP
+        Thread::ApartmentState state = Thread::AS_Unknown;        
+
+        if((!IsNilToken(tkEntryPoint)) && (TypeFromToken(tkEntryPoint) == mdtMethodDef)) {
+            if (scope->IsValidToken(tkEntryPoint))
+                state = SystemDomain::GetEntryPointThreadAptState(scope, tkEntryPoint);
+            else
+                ThrowHR(COR_E_BADIMAGEFORMAT);
+        }
+
+        // If the entry point has an explicit thread apartment state, set it
+        // before running the AppDomainManager initialization code.
+        if (state == Thread::AS_InSTA || state == Thread::AS_InMTA)
+            SystemDomain::SetThreadAptState(scope, state);
```

Comment:
Could you please delete `scope` argument of the `SetThreadAptState` method and the .NET Framework 1.0 compat quirk from it.